### PR TITLE
SplitBatchRunner

### PR DESCRIPTION
**Pull request by emailweixu**
Sometime we want to use a large batch for training but the memory is not enough. In this kind of situation, we can split the large batch into smaller batches and use lean_function to run the small batches. This can reduces the memory usage because the intermediate tensors will all be released.

So SplitBatchRunner is implemented to make this process very simple.

### CODE REVIEW

File: alf/utils/lean_function.py [hunk 1]
```diff
@@ -301,3 +301,56 @@ def _infer_device_type(*args):
...
+        batch_size = alf.nest.get_nest_batch_size((args, kwargs))
+        if self._max_batch_size <= 0 or batch_size <= self._max_batch_size:
+            return self._model._original_forward_for_lean_function(
+                *args, **kwargs, **non_batched_inputs)
+
+        outputs = []
+        for i in range(0, batch_size, self._max_batch_size):
+            batch_args, batch_kwargs = alf.nest.map_structure(
+                lambda x: x[i:i + self._max_batch_size], (args, kwargs))
+            outputs.append(
```
Comment (qiang-sean-liu) [inline alf/utils/lean_function.py:352]: are we supposed to have the same results from the full batch and the split batch? what if there are dropouts and batch norm in the model?

Comment (emailweixu) [reply alf/utils/lean_function.py:352]: As the doc-string notes, dropout is not supported for training.
BatchNorm will also be affected. But if the batch_size is not too small, its effect on training should be small.

File: alf/utils/lean_function.py [hunk 2]
```diff
@@ -301,3 +301,56 @@ def _infer_device_type(*args):
...
+    Example:
+
+    .. code-block:: python
+
+            model = SplitBatchRunner(model, max_batch_size=256)
+            output = model(input, non_batched_inputs={'mask': mask})
+
+    Args:
+        model (nn.Module): the model to run
+        max_batch_size (int): the maximum batch size to run the model.
```
Comment (hnyu) [inline alf/utils/lean_function.py]: Need to add a comment that if max_batch_size <=0, what will happen.

File: alf/utils/lean_function.py [hunk 3]
```diff
@@ -301,3 +301,56 @@ def _infer_device_type(*args):
...
+        outputs = []
+        for i in range(0, batch_size, self._max_batch_size):
+            batch_args, batch_kwargs = alf.nest.map_structure(
+                lambda x: x[i:i + self._max_batch_size], (args, kwargs))
+            outputs.append(
+                self._model(*batch_args, **batch_kwargs, **non_batched_inputs))
+
+        return alf.nest.map_structure(lambda *x: torch.cat(x, dim=0), *outputs)
+
+    def original_forward(self, *args, **kwargs):
```
Comment (hnyu) [inline alf/utils/lean_function.py:357]: Where is this function used?

File: alf/utils/lean_function.py [hunk 4]
```diff
@@ -301,3 +301,56 @@ def _infer_device_type(*args):
         return "cuda"
     else:
         return device_types[0]
+
+
+class SplitBatchRunner(torch.nn.Module):
+    """Split the input into smaller batches and run the model on each batch.
```
Comment (hnyu) [inline alf/utils/lean_function.py:307]: Need to clarify that lean_function will be used.

Comment (emailweixu) [reply alf/utils/lean_function.py:307]: comment added

File: alf/utils/lean_function.py (hunk 2 above)
Comment (emailweixu) [reply alf/utils/lean_function.py]: added

File: alf/utils/lean_function.py (hunk 1 above)
Comment (emailweixu) [reply alf/utils/lean_function.py:352]: added notes

File: alf/utils/lean_function.py (hunk 3 above)
Comment (emailweixu) [reply alf/utils/lean_function.py:357]: This is used by user of this class.

**Review by hnyu — APPROVED** [review]
